pkg/server/endpoints/endpoints.go

```diff
@@ -289,65 +291,141 @@ func (e *Endpoints) runLocalAccess(ctx context.Context, server *grpc.Server) err
+func (e *Endpoints) buildServerSpiffeAuthenticationFunction(bundleSet *x509bundle.Set) func(_ [][]byte, _ [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		if rawCerts == nil {
+			return nil
+		}
+
+		peerSpiffeID, _, err := x509svid.ParseAndVerify(rawCerts, bundleSet)
+		if err != nil {
+			return err
+		}
+
+		permissiveIDsSet := make(map[spiffeid.ID]struct{})
+		for _, adminID := range e.AdminIDs {
+			permissiveIDsSet[adminID] = struct{}{}
+		}
+
+		if !peerSpiffeID.MemberOf(e.TrustDomain) {
+			if _, ok := permissiveIDsSet[peerSpiffeID]; !ok {
+				return fmt.Errorf("unexpected ID %q", permissiveIDsSet)
+			}
+		}
+		return nil
+	}
+}
 
-			MinVersion: tls.VersionTLS12,
+// getBundleSource returns the bundle source to be used in the TLS peer verification
+func (e *Endpoints) getBundleSource(ctx context.Context) (bundleSet *x509bundle.Set, err error) {
+	bundleSet = x509bundle.NewSet()
 
-			NextProtos: []string{http2.NextProtoTLS},
-		}, nil
+	err = e.appendServerBundle(ctx, bundleSet)
+	if err != nil {
+		return
 	}
+
+	err = e.appendBundlesFromAdminIDs(ctx, bundleSet)
+	return
 }
 
-// getCerts queries the datastore and returns a TLS serving certificate(s) plus
-// the current CA root bundle.
-func (e *Endpoints) getCerts(ctx context.Context) ([]tls.Certificate, *x509.CertPool, error) {
-	bundle, err := e.DataStore.FetchBundle(dscache.WithCache(ctx), e.TrustDomain.IDString())
+// appendServerBundle appends the server bundle to the given bundle set.
+func (e *Endpoints) appendServerBundle(ctx context.Context, bundleSet *x509bundle.Set) error {
+	commonServerBundle, err := e.DataStore.FetchBundle(ctx, e.TrustDomain.IDString())
 	if err != nil {
-		return nil, nil, fmt.Errorf("get bundle from datastore: %w", err)
+		return fmt.Errorf("get bundle from datastore: %w", err)
 	}
-	if bundle == nil {
-		return nil, nil, errors.New("bundle not found")
+	if commonServerBundle == nil {
+		return fmt.Errorf("no bundle found for trust domain %s", e.TrustDomain.String())
+	}
+
+	serverBundle, err := parseBundle(e.TrustDomain, commonServerBundle)
+	if err != nil {
+		return fmt.Errorf("parse bundle: %w", err)
 	}
 
+	bundleSet.Add(serverBundle)
+	return nil
+}
+
+// appendBundlesFromAdminIDs appends the bundles from admin IDs trust domains to the given bundle set.
+func (e *Endpoints) appendBundlesFromAdminIDs(ctx context.Context, bundleSet *x509bundle.Set) error {
+	for _, adminID := range e.AdminIDs {
+		if !bundleSet.Has(adminID.TrustDomain()) {
+			commonBundle, err := e.DataStore.FetchBundle(ctx, adminID.TrustDomain().IDString())
+			if err != nil {
+				return fmt.Errorf("get bundle from datastore: %w", err)
+			}
+			if commonBundle == nil {
+				e.Log.
+					WithField(telemetry.AdminID, adminID.String()).
+					WithField(telemetry.TrustDomain, adminID.TrustDomain().String()).
+					Warn("No bundle found for foreign admin trust domain, admins from this trust doiman will not be able to connect")
```

Comment:
If an adminID from a foreign trust domain not yet federated with the spire server is configured, this warning will be triggered for all connection attempts. This could have been prevented if we had validation for admin_ids configuration during server start-up, what do you guys think about this? should I put work on another PR having this validation or this warning log (with the documentation) should be enough for an operator to understand and resolve this kind of misconfiguration?